activerecord/test/cases/calculations_test.rb

```diff
@@ -952,6 +952,89 @@ def test_group_by_attribute_with_custom_type
     assert_equal({ "proposed" => 2, "published" => 2 }, Book.group(:status).count)
   end
 
+  def test_select_avg_with_group_by_as_virtual_attribute_with_sql
+    rails_core = companies(:rails_core)
+
+    sql = <<~SQL
+      SELECT firm_id, AVG(credit_limit) AS avg_credit_limit
+      FROM accounts
+      WHERE firm_id = ?
+      GROUP BY firm_id
+    SQL
+
+    account = Account.find_by_sql([sql, rails_core]).first
+
+    # id was not selected, so it should be nil
+    # (cannot select id because it wasn't used in the GROUP BY clause)
+    assert_nil account.id
+
+    # firm_id was explicitly selected, so it should be present
+    assert_equal(rails_core, account.firm)
+
+    # avg_credit_limit should be present as a virtual attribute
+    assert_equal(52.5, account.avg_credit_limit)
+  end
+
+  def test_select_avg_with_group_by_as_virtual_attribute_with_ar
+    rails_core = companies(:rails_core)
+
+    account = Account
+      .select(:firm_id, "AVG(credit_limit) AS avg_credit_limit")
+      .where(firm: rails_core)
+      .group(:firm_id)
+      .order(:firm_id)
```

Comment:
Is this `.order(:firm_id)` necessary?

Reply:
Yes, because by default, `first!` adds an `ORDER BY id ASC` which doesn’t work, as `id` cannot be selected as it’s not part of the `GROUP BY` or an aggregate.

Reply:
I see. How about using `take` to no `ORDER BY` like as others SQL?

Reply:
Sure, I'll update it tomorrow. If the goal is to match the SQL, would you like me to add a `LIMIT 1` to the `find_by_sql`versions?

Reply:
Yeah, that is fine to me.